### PR TITLE
Make fingerprints configurable to prepare for a new default in the future

### PR DESCRIPTION
**Pull request by bdewater**
This is the continuation of the work by @medcat in #1708. I've rebased his work on master and made separate commits addressing changes requested in that pull request. If the maintainers of the gem give the green light I will squash the commits (preserving authorship) for clean commit history 😄

### CODE REVIEW

**Comment by bdewater** [comment]
@tute I've again rebased on master now that 5.0 is released. I assume it is too late now to change the default - following semantic versioning that would be 6.0. Could we work on getting the changes in with MD5 as default so that changing the default in the future is only a small change?

**Comment by tute** [comment]
> Could we work on getting the changes in with MD5 as default so that changing the default in the future is only a small change?

This makes perfect sense to me. Thank you!

**Comment by bdewater** [comment]
Rebased again @tute !

File: NEWS [hunk 1]
```diff
@@ -1,6 +1,22 @@
 master:
 * (port from 4.3) Improvement: the `URI adapter` now uses the content-disposition header to name the downloaded file.
 
+* Improvement: make the fingerprint digest configurable per attachment. The
+  default remains MD5 but this will change in a future version because it is
+  not considered secure anymore against intentional file corruption.
```
Comment (tute) [inline NEWS]: Can you gather resources on this assertion, so we provide more information to paperclip users?

File: README.md [hunk 2]
```diff
@@ -745,6 +745,16 @@ class AddAvatarFingerprintColumnToUser < ActiveRecord::Migration
 end
 ```
 
+The algorithm can be specified using a configuration option; it defaults to MD5.
```
Comment (tute) [inline README.md]: I'd add here "for backwards compatibility", so people know we are working on providing better defaults already.

File: lib/paperclip/io_adapters/abstract_adapter.rb [hunk 3]
```diff
@@ -8,8 +8,20 @@ class AbstractAdapter
...
 
+    def initialize(target, options = {})
+      @target = target
+      @options = options
+    end
+
     def fingerprint
-      @fingerprint ||= Digest::MD5.file(path).to_s
+      @fingerprint ||= begin
+        digest = @options.fetch(:hash_digest, Digest::MD5).new
```
Comment (tute) [inline lib/paperclip/io_adapters/abstract_adapter.rb]: Do we need to specify the default, if we already set it when paperclip starts?

File: lib/paperclip/io_adapters/identity_adapter.rb [hunk 4]
```diff
@@ -1,8 +1,11 @@
 module Paperclip
   class IdentityAdapter < AbstractAdapter
-    def new(adapter)
+    def new(adapter, _)
       adapter
     end
+
+    def initialize
```
Comment (tute) [inline lib/paperclip/io_adapters/identity_adapter.rb]: What does this definition with no arguments do?

File: lib/paperclip/io_adapters/uri_adapter.rb [hunk 5]
```diff
@@ -4,7 +4,8 @@ module Paperclip
   class UriAdapter < AbstractAdapter
     attr_writer :content_type
 
-    def initialize(target)
+    def initialize(target, options = {})
+      super
       @target = target
```
Comment (tute) [inline lib/paperclip/io_adapters/uri_adapter.rb]: `super` sets `@target`, we can drop this line.

File: lib/paperclip/tempfile_factory.rb [hunk 6]
```diff
@@ -13,7 +13,7 @@ def extension
     end
 
     def basename
-      Digest::MD5.hexdigest(File.basename(@name, extension))
+      Digest::SHA1.hexdigest(File.basename(@name, extension))
```
Comment (tute) [inline lib/paperclip/tempfile_factory.rb]: Should this be the configuration option, or be rollbacked to `MD5` for backwards compatibility?

File: lib/tasks/paperclip.rake [hunk 7]
```diff
@@ -64,7 +64,8 @@ namespace :paperclip do
       names = Paperclip::Task.obtain_attachments(klass)
       names.each do |name|
         Paperclip.each_instance_with_attachment(klass, name) do |instance|
-          if file = Paperclip.io_adapters.for(instance.send(name))
+          attachment = instance.send(name)
+          if file = Paperclip.io_adapters.for(instance.send(name), attachment.options[:adapter_options])
```
Comment (tute) [inline lib/tasks/paperclip.rake]: We can use `attachment` instead of `instance.send(name)` as first argument.

File: lib/tasks/paperclip.rake [hunk 8]
```diff
@@ -109,7 +123,20 @@ namespace :paperclip do
     end
   end
 
- desc "find missing attachments. Useful to know which attachments are broken"
+  desc "Upgrades the MD5 fingerprints to the SHA1 fingerprints."
```
Comment (tute) [inline lib/tasks/paperclip.rake]: Do you mean `SHA256` instead of `SHA1`? In any case, where do you set the option?

File: spec/paperclip/io_adapters/attachment_adapter_spec.rb [hunk 9]
```diff
@@ -41,8 +42,8 @@
       assert ! @subject.nil?
     end
 
-    it "generates a MD5 hash of the contents" do
-      expected = Digest::MD5.file(@file.path).to_s
+    it "generates a SHA1 hash of the contents" do
+      expected = Digest::SHA1.file(@file.path).to_s
```
Comment (tute) [inline spec/paperclip/io_adapters/attachment_adapter_spec.rb]: `SHA256`?

**Comment by tute** [comment]
This PR looks great, thank you for your great work, @bdewater! Left a few comments. Can you please address the ones you agree with on different commits, so as to ease the code review?

I might not be getting something around the `SHA1` vs `SHA256`. In that case, please explain why we have both.

Thank you!

File: lib/paperclip/io_adapters/identity_adapter.rb (hunk 4 above)
Comment (bdewater) [inline lib/paperclip/io_adapters/identity_adapter.rb]: I'll have to look at this closer later because at first glance I don't get what this class does. Changing it consistently breaks "Paperclip::Storage::S3 An attachment that uses S3 for storage and has a question mark in file name returns a replaced version for {path|url}" specs.

File: lib/paperclip/io_adapters/abstract_adapter.rb (hunk 3 above)
Comment (bdewater) [inline lib/paperclip/io_adapters/abstract_adapter.rb]: I've removed the duplication from Paperclip::Attachment.

File: lib/paperclip/attachment.rb [hunk 10]
```diff
@@ -33,7 +33,7 @@ def self.default_options
         :use_timestamp         => true,
         :whiny                 => Paperclip.options[:whiny] || Paperclip.options[:whiny_thumbnails],
         :validate_media_type   => true,
-        :adapter_options       => { hash_digest: Digest::MD5 },
+        :adapter_options       => { },
```
Comment (tute) [inline lib/paperclip/attachment.rb]: This hash contains all the default options. For documentation purposes, I'd leave the default here, and not in code as it was.

File: lib/paperclip/io_adapters/abstract_adapter.rb (hunk 3 above)
Comment (tute) [inline lib/paperclip/io_adapters/abstract_adapter.rb]: > I've removed the duplication from Paperclip::Attachment.

Let's remove this instance, and leave the one in the defaults hash for better code documentation. Thank you! 😃

File: spec/paperclip/io_adapters/data_uri_adapter_spec.rb [hunk 11]
```diff
@@ -20,7 +20,7 @@
   context "a new instance" do
     before do
       @contents = "data:image/png;base64,#{original_base64_content}"
-      @subject = Paperclip.io_adapters.for(@contents)
+      @subject = Paperclip.io_adapters.for(@contents, hash_digest: Digest::SHA1)
```
Comment (tute) [inline spec/paperclip/io_adapters/data_uri_adapter_spec.rb]: Should we use MD5 or SHA256 here?

**Comment by tute** [comment]
Thanks for your fast and great work, @bdewater! Left a comment around where to put the default hash function. Also, why do we mention SHA1 in the tests? Should we use the previous, or the new one?

**Comment by bdewater** [comment]
When moving the default hash location those S3 specs break in the same way as with the identity adapter 😠 I'll try some more in the weekend.

**Comment by bdewater** [comment]
I believe I've addressed all your comments @tute, I'll squash and rebase again if you agree it's ready for merging.

**Comment by tute** [comment]
Left one question. Please do squash and rebase. Thank you so much!

File: lib/paperclip/io_adapters/identity_adapter.rb [hunk 12]
```diff
@@ -1,12 +1,12 @@
 module Paperclip
   class IdentityAdapter < AbstractAdapter
-    def new(adapter)
+    def self.new(adapter, _)
```
Comment (tute) [inline lib/paperclip/io_adapters/identity_adapter.rb]: Why do we call this `adapter`, but `target` in other places?

**Comment by tute** [comment]
Looking good. After we fix `IdentityAdapter` and make it look like the other adapters we can merge. Thanks a ton. :)

**Comment by bdewater** [comment]
Done :)!

File: lib/paperclip/io_adapters/identity_adapter.rb [hunk 13]
```diff
@@ -1,12 +1,12 @@
 module Paperclip
   class IdentityAdapter < AbstractAdapter
-    def new(adapter)
-      adapter
```
Comment (tute) [inline lib/paperclip/io_adapters/identity_adapter.rb]: This is very confusing to me. Do we need this identity adapter? Do we need it to have a `new` class or instance method? The current PR changes how the Ruby language works (`.new` initializes and returns an instance), and that is not right. What happens if we drop this adapter?

Comment (bdewater) [inline lib/paperclip/io_adapters/identity_adapter.rb]: Perhaps @jyurek could shine some light on what this adapter is supposed to do because it confused me as well.

Comment (bdewater) [inline lib/paperclip/io_adapters/identity_adapter.rb]: I've reverted overriding `#new` in favor of the original implementation (overriding initialize to take no arguments and new as an instance method), I agree that it is less confusing. I've poked around a bit but it seems that one would need to refactor a lot to get rid of this adapter. That seems out of scope for this PR.

**Comment by tute** [comment]
This is a great change. Thanks so much for all your work! 🎉
